integration/test/Test/Federation.hs

```diff
@@ -0,0 +1,182 @@
+      newMsgNotif <- awaitNotification downUser1 downClient1 noValue 5 isNewMessageNotif
+      newMsgNotif %. "payload.0.qualified_conversation" `shouldMatch` objQidObject upBackendConv
+      newMsgNotif %. "payload.0.data.text" `shouldMatchBase64` "success message for down user"
+
+      -- FUTUREWORK: Uncomment after fixing this bug: https://wearezeta.atlassian.net/browse/WPB-3664
+      -- memberJoinNotif <- awaitNotification downUser1 downClient1 (Just newMsgNotif) 1 isMemberJoinNotif
+      -- memberJoinNotif %. "payload.0.qualified_conversation" `shouldMatch` objQidObject upBackendConv
+      -- asListOf objQidObject (memberJoinNotif %. "payload.0.data.users") `shouldMatch` mapM objQidObject [downUser2]
+
+      let isDelUserLeaveDownConvNotif =
+            allPreds
+              [ isConvLeaveNotif,
+                isNotifConv downBackendConv,
+                isNotifForUser delUser
+              ]
+      void $ awaitNotification downUser1 downClient1 (Just newMsgNotif) 1 isDelUserLeaveDownConvNotif
+
+      -- FUTUREWORK: Uncomment after fixing this bug: https://wearezeta.atlassian.net/browse/WPB-3664
+      -- void $ awaitNotification downUser1 downClient1 (Just newMsgNotif) 1 (allPreds [isConvLeaveNotif, isNotifConv upBackendConv, isNotifForUser otherUser])
+      -- void $ awaitNotification downUser1 downClient1 (Just newMsgNotif) 1 (allPreds [isConvLeaveNotif, isNotifConv upBackendConv, isNotifForUser delUser])
+
+      delUserDeletedNotif <- nPayload $ awaitNotification downUser1 downClient1 (Just newMsgNotif) 1 isDeleteUserNotif
+      objQid delUserDeletedNotif `shouldMatch` objQid delUser
+
+allPreds :: (Applicative f) => [a -> f Bool] -> a -> f Bool
+allPreds [] _ = pure True
+allPreds [p] x = p x
+allPreds (p1 : ps) x = (&&) <$> p1 x <*> allPreds ps x
+
+isDeleteUserNotif :: MakesValue a => a -> App Bool
+isDeleteUserNotif n =
+  nPayload n %. "type" `isEqual` "user.delete"
+
+isNewMessageNotif :: MakesValue a => a -> App Bool
+isNewMessageNotif n = fieldEquals n "payload.0.type" "conversation.otr-message-add"
+
+isMemberJoinNotif :: MakesValue a => a -> App Bool
+isMemberJoinNotif n = fieldEquals n "payload.0.type" "conversation.member-join"
+
+isConvLeaveNotif :: MakesValue a => a -> App Bool
+isConvLeaveNotif n = fieldEquals n "payload.0.type" "conversation.member-leave"
+
+isNotifConv :: (MakesValue conv, MakesValue a) => conv -> a -> App Bool
+isNotifConv conv n = fieldEquals n "payload.0.qualified_conversation" (objQidObject conv)
+
+isNotifForUser :: (MakesValue user, MakesValue a) => user -> a -> App Bool
+isNotifForUser user n = fieldEquals n "payload.0.data.qualified_user_ids.0" (objQidObject user)
+
+fieldEquals :: (MakesValue a, MakesValue b) => a -> String -> b -> App Bool
+fieldEquals a fieldSelector b = do
+  ma <- lookupField a fieldSelector `catchAll` const (pure Nothing)
+  case ma of
+    Nothing -> pure False
+    Just f ->
+      f `isEqual` b
```

Comment:
These general utilities should be in a separate module, shouldn't they?

Reply:
Good idea, moved.